File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/DeepLinkUtils.java
package com.garpr.android.misc;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.garpr.android.models.Endpoint;
import com.garpr.android.models.Region;
import com.garpr.android.models.RegionsBundle;

public interface DeepLinkUtils {

    @Nullable
    Intent[] buildIntentStack(@NonNull final Context context, @Nullable final Intent intent,
            @NonNull final Region region);

    @Nullable
    Intent[] buildIntentStack(@NonNull final Context context, @Nullable final String uri,
            @NonNull final Region region);

    @Nullable
    Intent[] buildIntentStack(@NonNull final Context context, @Nullable final Uri uri,
            @NonNull final Region region);

    @Nullable
    Endpoint getEndpoint(@Nullable final Intent intent);

    @Nullable
    Endpoint getEndpoint(@Nullable final String uri);

    @Nullable
    Endpoint getEndpoint(@Nullable final Uri uri);

    @Nullable
    Region getRegion(@Nullable final RegionsBundle regionsBundle, @Nullable final Endpoint endpoint);

    boolean isValidUri(@Nullable final Intent intent);

    boolean isValidUri(@Nullable final String uri);

    boolean isValidUri(@Nullable final Uri uri);

}

File: smash-ranks-android/app/src/main/java/com/garpr/android/misc/DeepLinkUtils.kt
package com.garpr.android.misc

import android.content.Context
import android.content.Intent
import android.net.Uri

import com.garpr.android.models.Endpoint
import com.garpr.android.models.Region
import com.garpr.android.models.RegionsBundle

interface DeepLinkUtils {

    fun buildIntentStack(context: Context, intent: Intent?, region: Region): List<Intent>?

    fun buildIntentStack(context: Context, uri: String?, region: Region): List<Intent>?

    fun buildIntentStack(context: Context, uri: Uri?, region: Region): List<Intent>?

    fun getEndpoint(intent: Intent?): Endpoint?

    fun getEndpoint(uri: String?): Endpoint?

    fun getEndpoint(uri: Uri?): Endpoint?

    fun getRegion(intent: Intent?, regionsBundle: RegionsBundle?): Region?

    fun getRegion(uri: String?, regionsBundle: RegionsBundle?): Region?

    fun getRegion(uri: Uri?, regionsBundle: RegionsBundle?): Region?

    fun isValidUri(intent: Intent?): Boolean

    fun isValidUri(uri: String?): Boolean

    fun isValidUri(uri: Uri?): Boolean

}
